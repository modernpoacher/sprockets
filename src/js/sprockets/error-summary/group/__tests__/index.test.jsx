import React from 'react'
import {
  snapshotOf
} from 'react-component-snapshot'
import classnames from 'classnames'

import '@testing-library/jest-dom'

import {
  render
} from '@testing-library/react'

import {
  getInstanceFromContainerElement
} from 'react-component-instance'

import Super from '#sprockets/super/components/group'
import Group from '#sprockets/sprockets/error-summary/group'

import {
  MOCK_ERROR_SUMMARY,
  MOCK_CHANGED_ERROR_SUMMARY
} from './definitions.mjs'

jest.mock('classnames', () => jest.fn().mockReturnValue('MOCK CLASSNAME'))

describe('#sprockets/sprockets/error-summary/group', () => {
  const MOCK_GROUP_REF = { current: null }

  describe('<Group />', () => {
    describe('With required props', () => {
      it('renders', () => {
        const {
          container: {
            firstElementChild: group
          }
        } = render(
          <Group />
        )

        expect(group)
          .toBeNull()
      })

      describe('Always', () => {
        it('invokes `getClassName`', () => {
          const getClassNameSpy = jest.spyOn(Group.prototype, 'getClassName')

          render(
            <Group />
          )

          expect(getClassNameSpy)
            .not.toHaveBeenCalled()
        })
      })

      /**
       *  Element is null
       */
      it('matches the snapshot', () => {
        const {
          container: {
            firstElementChild: group
          }
        } = render(
          <Group />
        )

        expect(snapshotOf(group))
          .toMatchSnapshot()
      })
    })

    describe('With additional props', () => {
      it('renders', () => {
        const {
          container: {
            firstElementChild: group
          }
        } = render(
          <Group
            errorSummary={MOCK_ERROR_SUMMARY}
          />
        )

        expect(group)
          .toBeInstanceOf(HTMLUListElement)
      })

      describe('Always', () => {
        it('invokes `getClassName`', () => {
          const getClassNameSpy = jest.spyOn(Group.prototype, 'getClassName')

          render(
            <Group
              errorSummary={MOCK_ERROR_SUMMARY}
            />
          )

          expect(getClassNameSpy)
            .toHaveBeenCalled()
        })
      })

      it('matches the snapshot', () => {
        const {
          container: {
            firstElementChild: group
          }
        } = render(
          <Group
            errorSummary={MOCK_ERROR_SUMMARY}
          />
        )

        expect(snapshotOf(group))
          .toMatchSnapshot()
      })
    })

    describe('`shouldComponentUpdate()`', () => {
      describe('`props` have changed', () => {
        describe('Prop `errorSummary` has changed', () => {
          it('returns true', () => {
            const {
              container
            } = render(
              <Group
                groupRef={MOCK_GROUP_REF}
                errorSummary={MOCK_ERROR_SUMMARY}
              />
            )

            /**
             *  @type {undefined | Group<GroupProps & CheckAnswersProps, GroupState & CheckAnswersState>}
             */
            const instance = getInstanceFromContainerElement(container)

            const {
              props,
              state
            } = instance

            expect(instance.shouldComponentUpdate(
              {
                ...props,
                errorSummary: MOCK_CHANGED_ERROR_SUMMARY
              },
              {
                ...state,
                errorSummary: MOCK_CHANGED_ERROR_SUMMARY
              }
            ))
              .toBe(true)
          })
        })
      })

      describe('`props` have not changed', () => {
        it('returns false', () => {
          const {
            container
          } = render(
            <Group
              groupRef={MOCK_GROUP_REF}
              errorSummary={MOCK_ERROR_SUMMARY}
            />
          )

          /**
           *  @type {undefined | Group<GroupProps & CheckAnswersProps, GroupState & CheckAnswersState>}
           */
          const instance = getInstanceFromContainerElement(container)

          const {
            props,
            state
          } = instance

          expect(instance.shouldComponentUpdate(
            {
              ...props,
              errorSummary: MOCK_ERROR_SUMMARY
            },
            {
              ...state,
              errorSummary: MOCK_ERROR_SUMMARY
            }
          ))
            .toBe(false)
        })
      })
    })

    describe('`getClassName()`', () => {
      it('invokes `classnames`', () => {
        /**
         *  Ensure `super.getClassName()` returns a value
         */
        const getClassNameSpy = jest.spyOn(Super.prototype, 'getClassName').mockReturnValue('MOCK CLASSNAME')

        const {
          container
        } = render(
          <Group />
        )

        const instance = getInstanceFromContainerElement(container)

        /**
         *  Ensure it is reset after render
         */
        classnames.mockClear()

        /**
         *  Ensure it is reset after render
         */
        getClassNameSpy.mockClear()

        instance.getClassName()

        expect(classnames)
          .toHaveBeenCalledWith('MOCK CLASSNAME', 'error-summary')
      })
    })
  })
})
