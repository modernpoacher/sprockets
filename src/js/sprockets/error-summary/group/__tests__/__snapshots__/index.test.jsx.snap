// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#sprockets/sprockets/error-summary/group <Group /> With additional props matches the snapshot 1`] = `
<ul
  className="MOCK CLASSNAME"
>
  <li
    className="error"
  >
    <a>
      <TextContent />
    </a>
  </li>
</ul>
`;

exports[`#sprockets/sprockets/error-summary/group <Group /> With required props matches the snapshot 1`] = `null`;
